helpers/testsuite/testing_helpers_test.go

```diff
@@ -0,0 +1,454 @@
+	chain, err := helpers.ParseCertificatesPEM(encodedCertChain)
+	checkError(err, t)
+
+	if len(chain) != len(requests) {
+		t.Log("Length of chain: " + strconv.Itoa(len(chain)))
+		t.Log("Length of requests: " + strconv.Itoa(len(requests)))
+		t.Fatal("Length of chain not equal to length of requests.")
+	}
+
+	mismatchOccurred := false
+	for i := 0; i < len(chain); i++ {
+		certEqualsRequest, unequalFields := certEqualsRequest(chain[i], requests[i])
+		if !certEqualsRequest {
+			mismatchOccurred = true
+			t.Log(
+				"Certificate " + strconv.Itoa(i) + " and request " +
+					strconv.Itoa(i) + " unequal.",
+			)
+			t.Log("Unequal fields for index " + strconv.Itoa(i) + ":")
+			for _, field := range unequalFields {
+				t.Log("\t" + field)
+			}
+		}
+	}
+
+	// TODO: check that each certificate is actually signed by the previous one
+
+	if mismatchOccurred {
+		t.Fatal("Unequal certificate(s) and request(s) found.")
+	}
+
+	// --- TEST: Create a chain of certificates with invalid path lengths. --- //
+
+	// Other invalid chains?
+}
+
+func TestCreateSelfSignedCert(t *testing.T) {
+
+	// --- TEST: Create a self-signed certificate from a CSR. --- //
+
+	// Make the request we will use to generate the certificate.
+	keyRequest := csr.KeyRequest{
+		Algo: "rsa",
+		Size: 2048,
+	}
+	CAConfig := csr.CAConfig{
+		PathLength: 1,
+		Expiry:     "1/1/2015",
+	}
+	request := csr.CertificateRequest{
+		CN: "example.com",
+		Names: []csr.Name{
+			{
+				C:  "US",
+				ST: "California",
+				L:  "San Francisco",
+				O:  "Internet Widgets, LLC",
+				OU: "Certificate Authority",
+			},
+		},
+		Hosts:      []string{"ca.example.com"},
+		KeyRequest: &keyRequest,
+		CA:         &CAConfig,
+	}
+
+	// Generate a self-signed certificate from the request.
+	encodedCertFromCode, _, err := CreateSelfSignedCert(request)
+	checkError(err, t)
+
+	// Now compare to a pre-made certificate made using a JSON file with the
+	// same request information. This JSON file is located in testdata/initCA
+	// and is called ca_csr.json.
+
+	CLIOutputFile := preMadeOutput
+	CLIOutput, err := ioutil.ReadFile(CLIOutputFile)
+	checkError(err, t)
+	encodedCertFromCLI := cleanCLIOutput(CLIOutput, "cert")
+
+	certFromCode, err := helpers.ParseSelfSignedCertificatePEM(encodedCertFromCode)
+	checkError(err, t)
+	certFromCLI, err := helpers.ParseSelfSignedCertificatePEM(encodedCertFromCLI)
+	checkError(err, t)
+
+	// Nullify any fields of the certificates which are dependent upon the time
+	// of the certificate's creation.
+	nullifyTimeDependency(certFromCode)
+	nullifyTimeDependency(certFromCLI)
+
+	if !reflect.DeepEqual(certFromCode, certFromCLI) {
+		unequalFields := checkFields(
+			*certFromCode, *certFromCLI, reflect.TypeOf(*certFromCode))
+		t.Log("The following fields were unequal:")
+		for _, field := range unequalFields {
+			t.Log(field)
+		}
+		t.Fatal("Certificates unequal.")
+	}
+
+}
+
+/* Compare two x509 certificate chains. We only compare relevant data to
```

Comment:
A style nit: we prefer the line commentary style over the C block-comment style.